src/java/org/apache/cassandra/index/sai/disk/IndexSearchResultIterator.java

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.exceptions.QueryCancelledException;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.index.sai.SSTableQueryContext;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.KeyRangeIterator;
+import org.apache.cassandra.index.sai.utils.KeyRangeUnionIterator;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.Throwables;
+
+public class IndexSearchResultIterator extends KeyRangeIterator
+{
+    private static final Logger logger = LoggerFactory.getLogger(IndexSearchResultIterator.class);
+
+    private final QueryContext context;
+    private final KeyRangeIterator union;
+    private final Set<SSTableIndex> referencedIndexes;
+
+    private IndexSearchResultIterator(KeyRangeIterator union, Set<SSTableIndex> referencedIndexes, QueryContext queryContext)
+    {
+        super(union.getMinimum(), union.getMaximum(), union.getCount());
+
+        this.union = union;
+        this.referencedIndexes = referencedIndexes;
+        this.context = queryContext;
+    }
+
+    /**
+     * Builds a new {@link IndexSearchResultIterator} that wraps a {@link KeyRangeUnionIterator} over the
+     * results of searching the {@link org.apache.cassandra.index.sai.memory.MemtableIndex} and the {@link SSTableIndex}es.
+     */
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public static IndexSearchResultIterator build(Expression expression,
+                                                  Set<SSTableIndex> sstableIndexes,
+                                                  AbstractBounds<PartitionPosition> keyRange,
+                                                  QueryContext queryContext)
+    {
+        List<KeyRangeIterator> subIterators = new ArrayList<>(1 + sstableIndexes.size());
+
+        KeyRangeIterator memtableIterator = expression.context.getMemtableIndexManager().searchMemtableIndexes(expression, keyRange);
+        if (memtableIterator != null)
+            subIterators.add(memtableIterator);
+
+        for (SSTableIndex sstableIndex : sstableIndexes)
+        {
+            try
+            {
+                queryContext.checkpoint();
+                queryContext.incSstablesHit();
+                assert !sstableIndex.isReleased() : "Index was released from the view during the query";
```

Comment:
nit: We might even want to upgrade this to an ISE

Reply:
Agreed, we can't continue in this case. I have also wrapped the error message (and other log and error messages) with `IndexContext.logMessage` so that they are clearer about which index the problem was with. It also makes them more consistent.